src/rust/engine/src/externs/interface.rs

```diff
@@ -610,13 +610,17 @@ py_class!(class PySession |py| {
           session_values: PyObject,
           cancellation_latch: PySessionCancellationLatch,
     ) -> CPyResult<Self> {
-      let session = Session::new(
-          scheduler.scheduler(py),
+      // NB: Session creation interacts with the Graph, which must not be accessed while the GIL is
+      // held.
+      let scheduler = scheduler.scheduler(py).clone();
+      let cancellation_latch = cancellation_latch.cancelled(py).clone();
+      let session = py.allow_threads(|| Session::new(
```

Comment:
Is there any way to ensure the "no GIL" condition for `Graph` access? Should `Graph` methods panic if the GIL is held?

Reply:
Hm... good idea. But I don't see a way to check whether the GIL is already acquired, and it's probably not worth adding thread local state for it.

Reply:
PyO3 exposes the necessary API:
- https://docs.rs/pyo3/0.14.5/pyo3/ffi/fn.PyGILState_Check.html
- https://docs.python.org/3/c-api/init.html#c.PyGILState_Check

Reply:
Good call. Yea, I think peppering a reusable debug assert around would be a great idea. I'll do that in a followup though, since it would be hard to cherry pick.